File: implementation/Scala/src/main/scala/synapses/model/encoding/Serialization.scala
package synapses.model.encoding

object Serialization {

  sealed trait Attribute

  final case class DiscreteAttribute(key: String,
                                     values: LazyList[String])
    extends Attribute

  case class ContinuousAttribute(key: String,
                                 min: Double,
                                 max: Double)
    extends Attribute with AttributeSerialized

  type Preprocessor = LazyList[Attribute]

  //

  sealed trait AttributeSerialized

  final case class DiscreteAttributeSerialized(key: String,
                                               values: List[String])
    extends AttributeSerialized

  type PreprocessorSerialized = List[AttributeSerialized]

}

File: implementation/Python/synapses/model/encoding/serialization.py
from typing import List

from dataclasses import dataclass
from functional.pipeline import Sequence


@dataclass(frozen=True)
class Attribute:
    key: str


@dataclass(frozen=True)
class DiscreteAttribute(Attribute):
    key: str
    values: Sequence


@dataclass(frozen=True)
class ContinuousAttribute(Attribute):
    key: str
    min: float
    max: float


Preprocessor = Sequence


##


@dataclass(frozen=True)
class AttributeSerialized:
    key: str


@dataclass(frozen=True)
class DiscreteAttributeSerialized(AttributeSerialized):
    key: str
    values: List[str]


@dataclass(frozen=True)
class ContinuousAttributeSerialized(AttributeSerialized):
    key: str
    min: float
    max: float
